aten/src/ATen/native/cpu/UpSampleKernel.cpp

```diff
@@ -0,0 +1,335 @@
+    auto input_data = (scalar_t*)data[1];
+
+    int64_t c = 0;
+    int64_t od = 0;
+    int64_t oh = 0;
+    int64_t ow = 0;
+    int64_t offset = output_data - output_data_base;
+    data_index_init(offset, c, channels, od, output_depth, oh, output_height, ow, output_width);
+
+    for (int64_t i = 0; i < n; i++) {
+      int64_t id = nearest_idx(od, input_depth, output_depth, scales[0]);
+      int64_t ih = nearest_idx(oh, input_height, output_height, scales[1]);
+      int64_t iw = nearest_idx(ow, input_width, output_width, scales[2]);
+      int64_t j = c * input_depth * input_height * input_width +
+                  id * input_height * input_width + ih * input_width + iw;
+      output_data[i] = input_data[j];
+      data_index_step(c, channels, od, output_depth, oh, output_height, ow, output_width);
+    }
+  };
+
+  std::vector<int64_t> strides(input_sizes.size(), 0);
+  auto input_expand = input.as_strided(output_sizes, strides);
+
+  auto iter = TensorIterator();
+  iter.set_check_mem_overlap(true);
+  iter.add_output(output);
+  iter.add_input(input_expand);
+  iter.build();
+
+  if (ndim == 3) {
+    // upsample nearest 1d
+    iter.for_each(loop1d);
+  } else if (ndim == 4) {
+    // upsample nearest 2d
+    iter.for_each(loop2d);
+  } else {
+    // upsample nearest 3d
+    iter.for_each(loop3d);
+  }
+
+  if (!output_.is_contiguous()) {
+    output_.copy_(output);
+  }
+}
+
+template <typename scalar_t, typename scale_type>
+void cpu_upsample_nearest_backward(
+    Tensor& grad_input_,
+    const Tensor& grad_output_,
+    const scale_type& scales) {
+  TORCH_CHECK(grad_input_.dtype() == grad_output_.dtype(), "expected dtype ", grad_output_.dtype(),
+              " for `grad_input` but got dtype ", grad_input_.dtype());
+  auto grad_output = grad_output_.contiguous();
+  auto grad_input = grad_input_.contiguous();
+
+  auto grad_output_data_base = grad_output.data_ptr<scalar_t>();
+  auto input_sizes = grad_input.sizes().vec();
+  auto output_sizes = grad_output.sizes().vec();
+  auto ndim = input_sizes.size();
+
+  // treat nbatch and channels as one dimension
+  int64_t channels = input_sizes[0] * input_sizes[1];
+  int64_t input_depth = (ndim == 5) ? input_sizes[2] : 0;
+  int64_t output_depth = (ndim == 5) ? output_sizes[2] : 0;
+  int64_t input_height = (ndim >= 4) ? input_sizes[ndim - 2] : 0;
+  int64_t output_height = (ndim >= 4) ? output_sizes[ndim - 2] : 0;
+  int64_t input_width = input_sizes[ndim - 1];
+  int64_t output_width = output_sizes[ndim - 1];
+
+  auto loop1d = [&](char** data, const int64_t* strides, int64_t n) {
+    auto grad_input_data = (scalar_t*)data[0];
+    auto grad_output_data = (scalar_t*)data[1];
+
+    int64_t c = 0;
+    int64_t ow = 0;
+    int64_t offset = grad_output_data - grad_output_data_base;
+    data_index_init(offset, c, channels, ow, output_width);
+
+    for (int64_t i = 0; i < n; i++) {
+      int64_t iw = nearest_idx(ow, input_width, output_width, scales[0]);
```

Comment:
**TODO:** this is not thread safe, have it fixed asap.

Reply:
**SOLVED** for the backward kernel, it is not thread safe to squash all dimensions and do parallelization. Since the parallel is done on `grad_output`, it is possible that multiple threads write a single address in `grad_input`.
Now the kernel does parallel on `batch_size` and `channel` dimension, thus multiple updates on a single address of `grad_input` will be carried out sequentially.